pkg/rtc/transport.go

```diff
@@ -981,138 +897,312 @@ func (t *PCTransport) configureReceiverDTX(enableDTX bool) {
 	}
 }
 
-func (t *PCTransport) CreateAndSendAnswer(enableDTX bool) error {
-	t.lock.RLock()
-	defer t.lock.RUnlock()
-
-	t.configureReceiverDTX(enableDTX)
+func (t *PCTransport) ICERestart() {
+	t.postEvent(event{
+		signal: signalICERestart,
+	})
+}
 
-	answer, err := t.pc.CreateAnswer(nil)
-	if err != nil {
-		return err
+func (t *PCTransport) OnStreamStateChange(f func(update *sfu.StreamStateUpdate) error) {
+	if t.streamAllocator == nil {
+		return
 	}
 
-	if t.preferTCP {
-		t.params.Logger.Infow("local answer (unfiltered)", "sdp", answer.SDP)
-	}
-	if err = t.pc.SetLocalDescription(answer); err != nil {
-		return err
+	t.streamAllocator.OnStreamStateChange(f)
+}
+
+func (t *PCTransport) AddTrackToStreamAllocator(subTrack types.SubscribedTrack) {
+	if t.streamAllocator == nil {
+		return
 	}
 
-	//
-	// Filter after setting local description as pion expects the answer
-	// to match between CreateAnswer and SetLocalDescription.
-	// Filtered answer is sent to remote so that remote does not
-	// see filtered candidates.
-	//
-	answer = t.filterCandidates(answer)
-	if t.preferTCP {
-		t.params.Logger.Infow("local answer (filtered)", "sdp", answer.SDP)
+	t.streamAllocator.AddTrack(subTrack.DownTrack(), sfu.AddTrackParams{
+		Source:      subTrack.MediaTrack().Source(),
+		IsSimulcast: subTrack.MediaTrack().IsSimulcast(),
+		PublisherID: subTrack.MediaTrack().PublisherID(),
+	})
+}
+
+func (t *PCTransport) RemoveTrackFromStreamAllocator(subTrack types.SubscribedTrack) {
+	if t.streamAllocator == nil {
+		return
 	}
 
-	if t.onAnswer != nil {
-		go t.onAnswer(answer)
+	t.streamAllocator.RemoveTrack(subTrack.DownTrack())
+}
+
+func (t *PCTransport) SetPreviousAnswer(answer *webrtc.SessionDescription) {
+	t.postEvent(event{
+		signal: signalPreviousAnswer,
+		data:   answer,
+	})
+}
+
+func (t *PCTransport) postEvent(event event) {
+	t.eventChMu.RLock()
+	if t.isClosed.Load() {
+		t.eventChMu.RUnlock()
+		return
+	}
+
+	select {
+	case t.eventCh <- event:
+	default:
+		t.params.Logger.Warnw("event queue full", nil)
+	}
+	t.eventChMu.RUnlock()
+}
+
+func (t *PCTransport) processEvents() {
+	for event := range t.eventCh {
+		err := t.handleEvent(&event)
+		if err != nil {
+			t.params.Logger.Errorw("error handling event", err, "event", event.String())
+			if t.onNegotiationFailed != nil {
+				t.onNegotiationFailed()
+			}
```

Comment:
Just consolidating all processing here. Basically, any error is not recoverable (or at least will lead to confusing state). So, checking this one place and forcing a full reconnect on any error.

Reply:
Wondering not all error is critical, if a ice candidate is not valid but others are valid, peerconnection might still connect.

Reply:
Good point @cnderrauber . But, I was thinking (actually more hoping :-) ) that in those there is no error returned by Pion for `AddICECandidate`. It just ignores it. Will check on it.

Reply:
@cnderrauber Checked this. Pion filters out all non-critical errors and returns nil. So, treating all errors as bad should be fine.